### PR TITLE
feat: Create node group user data from given template

### PR DESCRIPTION
**Pull request by scalen**
# PR o'clock

Closes https://github.com/terraform-aws-modules/terraform-aws-eks/issues/1644

## Description

Prior to this commit, user data for node groups was generated from a
prescribed template, and user data in other formats (such as the TOML
used to configure Bottlerocket instances, see link) was not supported.

This change allows a node_group to specify its own user data template
file, the template's extra arguments and the user data's mime type;
this in turn supports alternative forms of user data as required by any
given AMI.

https://github.com/bottlerocket-os/bottlerocket#using-user-data

### Checklist

- [x] README.md has been updated after any changes to variables and outputs. See https://github.com/terraform-aws-modules/terraform-aws-eks/#doc-generation

### CODE REVIEW

**Comment by stevehipwell** [comment]
@scalen I was planning on doing this work so I'm very happy to see you've already picked it up. Could you rebase this onto the latest master branch and I'll give it a review?

**Comment by stevehipwell** [comment]
@scalen I'm not sure there is a precedent for complex objects in the node group config so I'd suggest removing the nesting and using `user_data_` as a prefix (e.g. `user_data_mime_type`), there are only 3 variables. This will also make it possible to only override one of the values without having to specifically set the others to their defaults, which wouldn't be as easy to use and could cause outdated defaults to be used if they were ever changed. Other than that this looks really good. How much testing have you done?

**Comment by stevehipwell** [comment]
@scalen if you make the changes to the variables I'll review again and then we can see if @daroga0002 can approve running CI. Once this is merged I might take a look at adding the changes in my previous PRs and the changes here for launch template workers.

**Comment by scalen** [comment]
@stevehipwell Requested changes made in 7135ba9338f52513510152c22d7e3b256f931951

**Review by stevehipwell — APPROVED** [review]
Looks good to me, @daroga0002 what do you think?

**Comment by jaimehrubiks** [comment]
Does this modify existing node_groups with user data like these:

      pre_userdata = join("\n\n#----------------\n\n",[
        templatefile("${path.module}/templates/_volume_fg6001.sh", {}),
        templatefile("${path.module}/templates/_volume_fg6002.sh", {}),
      ])

If you don't change anything?

**Comment by scalen** [comment]
> Does this modify existing node_groups with user data like these:
> 
> ```
>   pre_userdata = join("\n\n#----------------\n\n",[
>     templatefile("${path.module}/templates/_volume_fg6001.sh", {}),
>     templatefile("${path.module}/templates/_volume_fg6002.sh", {}),
>   ])
> ```
> 
> If you don't change anything?

No, the default behaviour remains the same.

**Comment by bryantbiggs** [comment]
ah, good ol' userdata again. I think we just need to provide a bring-your-own user_data functionality. something like:

```hcl
	...
	user_data = var.create_user_data ?  base64encode(local.launch_template_userdata_rendered[count.index]) : var.user_data_base64
	...
```

thoughts @daroga0002 ?

**Comment by stevehipwell** [comment]
@bryantbiggs isn't this PR providing a BYO fallback implementation?

**Comment by bryantbiggs** [comment]
yes, but in a way that is getting very complex and convoluted

**Comment by stevehipwell** [comment]
@bryantbiggs it's a complex problem that's being solved. The templated solutions currently provided by the input variables have a pretty good UX and for everyone else this PR will allow them to own the full complexity themselves.

**Comment by scalen** [comment]
> yes, but in a way that is getting very complex and convoluted

It is a little convoluted, but by providing the template and additional args separately, it means we can utilise many variables created within the module. If we required the user data to be passed in whole, we may struggle to get some of the same synchronisation that is possible/trivial with the current approach.

File: modules/node_groups/locals.tf
Comment (bdwyertech) [inline modules/node_groups/locals.tf:38]: I think the lookup reference to `userdata_template_file` here was meant to be `user_data_template_extra_args`

Comment (scalen) [reply modules/node_groups/locals.tf:38]: You're right, this is incorrect, but [it should actually be `user data_template_extra_args`](https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/locals.tf#L255-L259).

**Review by bdwyertech — APPROVED** [review]

**Review by bdwyertech — APPROVED** [review]

**Review by djordje-petrovic — APPROVED** [review]

**Comment by stevehipwell** [comment]
@daroga0002 @antonbabenko could we get this reviewed and merged?

**Comment by stevehipwell** [comment]
@scalen do you not need to provide a variable to customise the `block_device_mappings.device_name` to support Bottlerocket (and other OSs)?

See https://github.com/terraform-aws-modules/terraform-aws-eks/pull/1695 & https://github.com/terraform-aws-modules/terraform-aws-eks/pull/1685.

**Comment by scalen** [comment]
> @scalen do you not need to provide a variable to customise the `block_device_mappings.device_name` to support Bottlerocket (and other OSs)?
> 
> See #1695 & #1685.

We haven't seen any issues running of this PR branch, but then we haven't specified any significant configuration.  I assume the suggestion is to make a more flexible version of [this change](https://github.com/terraform-aws-modules/terraform-aws-eks/pull/1695#issuecomment-975500387), to let the user define the volume to configure:

> `device_name = length(split("BOTTLEROCKET", each.value["ami_type"])) > 1 ? "/dev/xvdb" : "/dev/xvda"`

This still means only one volume will be configured, but it seems that will be fine for the Bottlerocket case 🤷🏼

**Comment by stevehipwell** [comment]
@scalen that is correct; I assume that if you tried to configure the volume that it'd be applied to the wrong Bottlerocket volume (as is discussed in the other PR). A generic way to pass in the block device name would be a good solution, as would adding it in to the Bottlerocket example.

**Comment by scalen** [comment]
> I made the original volume comment on #1695, saying that we shouldn't really care about the `xvda` volume. For context, `xvda` contains operating system itself: the active & passive partitions, the bootloader, the dm-verity data and etc. However, I can see the use case where you want to customize both `xvda` and `xvdb`. For example, in case you are required to encrypt all your ebs devices.

I thought of that too, but that is really starting to stray far from the point of the PR, which was to add support for alternative user data formats back to node_groups.  Making an extra value configurable in order to slightly extend the existing functionality to behave as one might expect for the specific Bottlerocket case is arguably related; I would also argue, however, that extending the module to support the configuration of multiple distinct volumes is a new feature, which should get its own PR.

**Comment by stevehipwell** [comment]
@hi-artem @scalen the Bottlerocket OS volume is readonly so I think that if this PR supports changing the device name we're good here. Adding support for additional devices should be a new PR, but I think we need to make sure that the right Bottlerocket (or other OS) device is being configured in this PR.

**Comment by scalen** [comment]
> I think we need to make sure that the right Bottlerocket (or other OS) device is being configured in this PR.

Done, @stevehipwell .

**Comment by FrediWeber** [comment]
> @hi-artem @scalen the Bottlerocket OS volume is readonly so I think that if this PR supports changing the device name we're good here. Adding support for additional devices should be a new PR, but I think we need to make sure that the right Bottlerocket (or other OS) device is being configured in this PR.

Is the Bottlerocket volume RO from an EC2 standpoint? I thought the volume just gets mounted as RO.

**Comment by stevehipwell** [comment]
@FrediWeber that's worth exploring. I guess the `block_device` variable could be made `block_devices` and take a list and then the mapping be dynamically looped which would support all potential cases? @scalen thoughts?

**Comment by FrediWeber** [comment]
We just configured a launch template for a classic ASG and needed to configure both block devices to be encrypted and use a custom KMS key to be compliant with our regulations. I think it would be beneficial if we could encrypt all block devices on the Bottlerocket instances.

**Comment by idan-gur** [comment]
What is the blockage on this PR?

**Comment by stevehipwell** [comment]
@daroga0002 @antonbabenko could you approve the workflow even if you don't have time to review yet?

**Comment by daroga0002** [comment]
> @daroga0002 @antonbabenko could you approve the workflow even if you don't have time to review yet?

I dont have permissions and @antonbabenko  is on re:Invent so probably he will be back next week.

I am a little stuck in last 2 weeks by some work, but I will try soon perform a review pending PRs.

**Comment by scalen** [comment]
Tests from before rebase to fix formatting: https://github.com/terraform-aws-modules/terraform-aws-eks/runs/4397905746

**Comment by lucasreed** [comment]
Checking in to see if this is in the pipeline for reviewing. Would love to see this released!

**Comment by antonbabenko** [comment]
This issue has been resolved in version 18.0.0 :tada:

**Pull request closed without merging**
